test/e2e/upgrades/apparmor.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrades
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	api "k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/test/e2e/common"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/onsi/gomega/gstruct"
+)
+
+// AppArmorUpgradeTest tests that AppArmor profiles are enforced & usable across upgrades.
+type AppArmorUpgradeTest struct {
+	pod *api.Pod
+}
+
+// Setup creates a secret and then verifies that a pod can consume it.
+func (t *AppArmorUpgradeTest) Setup(f *framework.Framework) {
+	common.SkipIfAppArmorNotSupported()
+	By("Loading AppArmor profiles to nodes")
+	common.LoadAppArmorProfiles(f)
+
+	// Create the initial test pod.
+	By("Creating a long-running AppArmor enabled pod.")
+	t.pod = common.CreateAppArmorTestPod(f, false)
+
+	// Verify initial state.
+	t.verifyNodesAppArmorEnabled(f)
+	t.verifyNewPodSucceeds(f)
+}
+
+// Test waits for the upgrade to complete, and then verifies that a
+// pod can still consume the secret.
+func (t *AppArmorUpgradeTest) Test(f *framework.Framework, done <-chan struct{}, upgrade UpgradeType) {
+	<-done
+	if upgrade == MasterUpgrade {
+		t.verifyPodStaysUp(f)
+	}
+	t.verifyNodesAppArmorEnabled(f)
+	t.verifyNewPodSucceeds(f)
+}
+
+// Teardown cleans up any remaining resources.
+func (t *AppArmorUpgradeTest) Teardown(f *framework.Framework) {
+	// rely on the namespace deletion to clean up everything
+	By("Logging container failures")
+	framework.LogFailedContainers(f.ClientSet, f.Namespace.Name, framework.Logf)
+}
+
+func (t *AppArmorUpgradeTest) verifyPodStaysUp(f *framework.Framework) {
```

Comment:
Yeah, that was actually the original intent, but then I realized the `Get` call would fail while the apiserver is restarting, so decided to just check after the upgrade was complete. Is there a better way to do this?

I'll rename for now.